src/App.tsx

```diff
@@ -1,26 +1,291 @@
+      inputRef.current?.focus();
+    }
+  };
+
+  const handleToggleTodo = async (todo: Todo) => {
+    setIdToLoad(prev => [...prev, todo.id]);
+
+    try {
+      const toggledTodo = await updateTodo(todo.id, {
+        completed: !todo.completed,
+      });
+
+      setTodos(currTodos =>
+        currTodos.map(currTodo => {
+          return currTodo.id === toggledTodo.id ? toggledTodo : currTodo;
+        }),
+      );
+    } catch {
+      setError('Unable to update a todo');
+    } finally {
+      setIdToLoad([]);
+    }
+  };
+
+  const handleToggleAll = async () => {
+    let todosToToggle = todos.filter(todo => !todo.completed);
+
+    if (todosToToggle.length === 0) {
+      setIdToLoad(todos.map(todo => todo.id));
+      todosToToggle = [...todos];
+    } else {
+      setIdToLoad(todosToToggle.map(todo => todo.id));
+    }
+
+    try {
+      await Promise.all(
+        todosToToggle.map(async todo => {
+          try {
+            const toggledTodo = await updateTodo(todo.id, {
+              completed: !todo.completed,
+            });
+
+            setTodos(prev =>
+              prev.map(currTodo => {
+                return currTodo.id === toggledTodo.id ? toggledTodo : currTodo;
+              }),
+            );
+          } catch {
+            setError('Unable to update a todo');
+          }
+        }),
+      );
+    } finally {
+      setIdToLoad([]);
+    }
+  };
+
+  const TodoTitleSubmit = async (todo: Todo) => {
+    const normalizedTitle = newTitle.trim();
+
+    if (todo.title === normalizedTitle) {
+      return;
+    }
+
+    setIdToLoad(prev => [...prev, todo.id]);
+
+    try {
+      const editedTodo = await updateTodo(todo.id, {
+        title: normalizedTitle,
+      });
+
+      setTodos(currTodos =>
+        currTodos.map(currTodo =>
+          currTodo.id === editedTodo.id ? editedTodo : currTodo,
+        ),
+      );
+    } catch {
+      setError('Unable to update a todo');
+
+      return;
+    } finally {
+      setIdToLoad([]);
+    }
+
+    setEditingTodo(null);
+    setNewTitle('');
+  };
+
+  const filteredTodos = useMemo(
+    () =>
+      todos.filter(todo => {
+        switch (filter) {
+          case FilterType.Active:
+            return !todo.completed;
+          case FilterType.Completed:
+            return todo.completed;
+          default:
+            return true;
+        }
+      }),
+    [todos, filter],
+  );
+
+  const loadTodos = async () => {
+    inputRef.current?.focus();
+    setError('');
+    setIsLoading(true);
+
+    try {
+      const fetchedTodos = await getTodos();
+
+      setTodos(fetchedTodos);
+    } catch {
+      setError('Unable to load todos');
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  const ErrorNotificationClear = () => {
+    let timeoutId: number;
+
+    if (error?.length) {
+      timeoutId = window.setTimeout(() => setError(null), 3000);
+    }
+
+    return () => {
+      clearTimeout(timeoutId);
+    };
+  };
```

Comment:
thats shouldn't be handler
you can manage clear timeout in useEffect in ErrorNotification component